streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java

```diff
@@ -0,0 +1,479 @@
+        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        streamsConfiguration.put(StreamsConfig.METRICS_RECORDING_LEVEL_CONFIG, Sensor.RecordingLevel.DEBUG.name);
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 10 * 1024 * 1024L);
+    }
+
+    @Before
+    public void after() throws InterruptedException {
+        CLUSTER.deleteTopics(STREAM_INPUT, STREAM_OUTPUT_1, STREAM_OUTPUT_2, STREAM_OUTPUT_3, STREAM_OUTPUT_4);
+    }
+
+    private void startApplication() throws Exception {
+
+        kafkaStreams = new KafkaStreams(builder.build(), streamsConfiguration);
+        kafkaStreams.start();
+
+        Thread.sleep(10000);
+    }
+
+    private void closeApplication() throws Exception {
+        Thread.sleep(10000);
+        kafkaStreams.close();
+        kafkaStreams.cleanUp();
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+    @Test
+    public void testStreamMetric() throws Exception {
+
+        stream = builder.stream(STREAM_INPUT, Consumed.with(Serdes.Integer(), Serdes.String()));
+        stream.to(STREAM_OUTPUT_1, Produced.with(Serdes.Integer(), Serdes.String()));
+        builder.table(STREAM_OUTPUT_1, Materialized.as(Stores.inMemoryKeyValueStore("myStoreInMemory")).withCachingEnabled())
+                .toStream()
+                .to(STREAM_OUTPUT_2);
+        builder.table(STREAM_OUTPUT_2, Materialized.as(Stores.persistentKeyValueStore("myStorePersistentKeyValue")).withCachingEnabled())
+                .toStream()
+                .to(STREAM_OUTPUT_3);
+        builder.table(STREAM_OUTPUT_3, Materialized.as(Stores.lruMap("myStoreLruMap", 10000)).withCachingEnabled())
+                .toStream()
+                .to(STREAM_OUTPUT_4);
+
+        startApplication();
+
+        // metric level : Thread
+        testThreadMetric();
+
+        // metric level : Task
+        testTaskMetric();
+
+        // metric level : Processor
+        testProcessorMetric();
+
+        // metric level : Store (in-memory-state, in-memory-lru-state, rocksdb-state)
+        testStoreMetricByType(STREAM_STORE_IN_MEMORY_STATE_METRICS);
+        testStoreMetricByType(STREAM_STORE_IN_MEMORY_LRU_STATE_METRICS);
+        testStoreMetricByType(STREAM_STORE_ROCKSDB_STATE_METRICS);
+
+        //metric level : Cache
+        testCacheMetric();
+
+        closeApplication();
+
+        // check all metrics de-registered
+        final List<Metric> listMetricAfterClosingApp = new ArrayList<Metric>(kafkaStreams.metrics().values()).stream().filter(m -> m.metricName().group().contains(STREAM_STRING)).collect(Collectors.toList());
+        Assert.assertEquals(0, listMetricAfterClosingApp.size());
+
+    }
+
+    @Test
+    public void testStreamMetricOfWindowStore() throws Exception {
```

Comment:
I was originally thinking about getting the list of all metrics from kafkaStreams.metrics() which contains producer, consumer, admin and streams' own metrics (thread, task, processor node, store, cache), and check that they all exist with the exact number of metrics (with the PR for KIP-414 is in it should be easy to get the corresponding client id for different modules).

But after reading @vvcephei 's comment I think I'm convinced that we can save on getting non-streams embedded client's metrics later, and also the actual metrics name validation may better not in Streams metrics test, so we'd probably only check that the corresponding groups with clientIds existed rather than checking each metric exists. So I'm fine with the current scope of this PR.